File: mobile/lib/shared/models/message.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'user.dart';

part 'message.freezed.dart';
part 'message.g.dart';

enum ConversationType { direct, group }

enum MessageType { text, image, video, audio, file, sticker, system }

enum MessageStatus { sending, sent, delivered, read, failed }

@freezed
class Conversation with _$Conversation {
  const factory Conversation({
    required String id,
    required ConversationType type,
    String? name,
    String? avatarUrl,
    required String createdBy,
    DateTime? lastMessageAt,
    required DateTime createdAt,
    required DateTime updatedAt,
    @Default([]) List<Participant> participants,
    Message? lastMessage,
  }) = _Conversation;

  factory Conversation.fromJson(Map<String, dynamic> json) =>
      _$ConversationFromJson(json);
}

@freezed
class Participant with _$Participant {
  const factory Participant({
    required String id,
    required String conversationId,
    required String userId,
    @Default('member') String role,
    required DateTime joinedAt,
    DateTime? leftAt,
    DateTime? mutedUntil,
    User? user,
  }) = _Participant;

  factory Participant.fromJson(Map<String, dynamic> json) =>
      _$ParticipantFromJson(json);
}

@freezed
class Message with _$Message {
  const factory Message({
    required String id,
    required String conversationId,
    required String senderId,
    required MessageType type,
    required List<int> content, // Encrypted content
    String? stickerId,
    String? replyToId,
    @Default(MessageStatus.sending) MessageStatus status,
    required DateTime createdAt,
    required DateTime updatedAt,
    DateTime? deletedAt,
    User? sender,
    Message? replyTo,
    @Default([]) List<Receipt> receipts,
  }) = _Message;

  factory Message.fromJson(Map<String, dynamic> json) => _$MessageFromJson(json);
}

@freezed
class Receipt with _$Receipt {
  const factory Receipt({
    required String id,
    required String messageId,
    required String userId,
    required String type, // 'delivered' or 'read'
    required DateTime createdAt,
  }) = _Receipt;

  factory Receipt.fromJson(Map<String, dynamic> json) => _$ReceiptFromJson(json);
}

// Decrypted message content
@freezed
class DecryptedContent with _$DecryptedContent {
  const factory DecryptedContent.text({
    required String text,
  }) = TextContent;

  const factory DecryptedContent.image({
    required String url,
    String? caption,
    int? width,
    int? height,
  }) = ImageContent;

  const factory DecryptedContent.file({
    required String url,
    required String fileName,
    required int fileSize,
    String? mimeType,
  }) = FileContent;

  const factory DecryptedContent.sticker({
    required String stickerId,
    required String packId,
    required String imageUrl,
  }) = StickerContent;
}

File: backend-rs/src/models/message.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Message {
    pub id: Uuid,
    pub conversation_id: Uuid,
    pub sender_id: Uuid,
    #[serde(rename = "type")]
    pub message_type: MessageType,
    pub content: Vec<u8>,
    pub sticker_id: Option<Uuid>,
    pub reply_to_id: Option<Uuid>,
    pub status: MessageStatus,
    pub edited_at: Option<DateTime<Utc>>,
    pub deleted_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "message_type", rename_all = "lowercase")]
#[serde(rename_all = "lowercase")]
pub enum MessageType {
    Text,
    Image,
    Video,
    Audio,
    File,
    Sticker,
    System,
}

impl Default for MessageType {
    fn default() -> Self {
        Self::Text
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "message_status", rename_all = "lowercase")]
#[serde(rename_all = "lowercase")]
pub enum MessageStatus {
    Sending,
    Sent,
    Delivered,
    Read,
    Failed,
}

impl Default for MessageStatus {
    fn default() -> Self {
        Self::Sent
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Receipt {
    pub id: Uuid,
    pub message_id: Uuid,
    pub user_id: Uuid,
    #[serde(rename = "type")]
    pub receipt_type: ReceiptType,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "receipt_type", rename_all = "lowercase")]
#[serde(rename_all = "lowercase")]
pub enum ReceiptType {
    Delivered,
    Read,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageWithSender {
    #[serde(flatten)]
    pub message: Message,
    pub sender: Option<super::User>,
}
